iceoryx_posh/include/iceoryx_posh/internal/runtime/posh_runtime_impl.hpp

```diff
@@ -0,0 +1,129 @@
+#include "iceoryx_posh/runtime/posh_runtime.hpp"
+
+#include <mutex>
+
+namespace iox
+{
+namespace runtime
+{
+enum class RuntimeLocation
+{
+    SEPARATE_PROCESS_FROM_ROUDI,
+    SAME_PROCESS_LIKE_ROUDI,
+};
+
+/// @brief The runtime that is needed for each application to communicate with the RouDi daemon
+class PoshRuntimeImpl : public PoshRuntime
+{
+  public:
+    PoshRuntimeImpl(const PoshRuntimeImpl&) = delete;
+    PoshRuntimeImpl& operator=(const PoshRuntimeImpl&) = delete;
+    PoshRuntimeImpl(PoshRuntimeImpl&&) = delete;
+    PoshRuntimeImpl& operator=(PoshRuntimeImpl&&) = delete;
+    virtual ~PoshRuntimeImpl() noexcept;
+
+    /// @copydoc PoshRuntime::findService
+    cxx::expected<InstanceContainer, FindServiceError>
+    findService(const capro::ServiceDescription& serviceDescription) noexcept override;
+
+    /// @copydoc PoshRuntime::offerService
+    bool offerService(const capro::ServiceDescription& serviceDescription) noexcept override;
+
+    /// @copydoc PoshRuntime::stopOfferService
+    void stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept override;
+
+    /// @copydoc PoshRuntime::getMiddlewarePublisher
+    PublisherPortUserType::MemberType_t*
+    getMiddlewarePublisher(const capro::ServiceDescription& service,
+                           const popo::PublisherOptions& publisherOptions = popo::PublisherOptions(),
+                           const PortConfigInfo& portConfigInfo = PortConfigInfo()) noexcept override;
+
+    /// @copydoc PoshRuntime::getMiddlewareSubscriber
+    SubscriberPortUserType::MemberType_t*
+    getMiddlewareSubscriber(const capro::ServiceDescription& service,
+                            const popo::SubscriberOptions& subscriberOptions = popo::SubscriberOptions(),
+                            const PortConfigInfo& portConfigInfo = PortConfigInfo()) noexcept override;
+
+    /// @copydoc PoshRuntime::getMiddlewareInterface
+    popo::InterfacePortData* getMiddlewareInterface(const capro::Interfaces interface,
+                                                    const NodeName_t& nodeName = {""}) noexcept override;
+
+    /// @copydoc PoshRuntime::getMiddlewareApplication
+    popo::ApplicationPortData* getMiddlewareApplication() noexcept override;
+
+    /// @copydoc PoshRuntime::getMiddlewareConditionVariable
+    popo::ConditionVariableData* getMiddlewareConditionVariable() noexcept override;
+
+    /// @copydoc PoshRuntime::createNode
+    NodeData* createNode(const NodeProperty& nodeProperty) noexcept override;
+
+    /// @copydoc PoshRuntime::getServiceRegistryChangeCounter
+    const std::atomic<uint64_t>* getServiceRegistryChangeCounter() noexcept override;
+
+    /// @copydoc PoshRuntime::sendRequestToRouDi
+    bool sendRequestToRouDi(const IpcMessage& msg, IpcMessage& answer) noexcept override;
+
+  protected:
+    friend class PoshRuntime;
+    friend class roudi::RuntimeTestInterface;
+
+    // Protected constructor for IPC setup
```

Comment:
You could remove this comment as `protected` and `private` methods are usually not documented or convert it to doxygen.

Reply:
These are not doxygen comments and will not be rendered to the documentation. Sometimes it's helpful to have comments on non public methods to explain the reasoning why it's not public.